core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala

```diff
@@ -204,9 +207,33 @@ class ShardingContainerPoolBalancer(
     MetricEmitter.emitGaugeMetric(OFFLINE_INVOKER_BLACKBOX, schedulingState.blackboxInvokers.count(_.status == Offline))
   }
 
+  class RecentInvokerQueue {
+    private val maxRecentInvokers = lbConfig.maxRecentInvokers
+    private var actionLastInvokers: Queue[(Int, Long)] = Queue.empty[(Int, Long)]
+
+    def enqueue(invoker: Int): Unit = {
+      actionLastInvokers = actionLastInvokers.enqueue((invoker, Instant.now.toEpochMilli))
+      while (actionLastInvokers.size > maxRecentInvokers) { actionLastInvokers = actionLastInvokers.dequeue._2 }
+    }
+
+    /** current algorithm just orders by most frequent the last x invokers completed activations for an action */
+    def getPreferredInvokers: Seq[Int] = {
+      //filter out invokers that would represent an activation over 10 minutes old
+      val validActionLastInvokers =
+        actionLastInvokers.filter(invokerWithTimestamp => Instant.now.toEpochMilli - invokerWithTimestamp._2 < 600000)
```

Comment:
Should the 10 minute value be configurable?

Reply:
Yea it should be equal to the config for the invoker of idle time between activations before killing the container. That config isn't accessible in the controller right now so I just hardcoded it to what we use for that value which is the default until we test it.

Reply:
nit:
Not sure of the code style for this project. I find named functions or constants helpful to communicate the purpose of magic numbers. Not sure if there are downstream issues to that.

It would make it easier for someone else to identify where code needs to change once this could be configured